### DISCUSSION TITLE
[Feature Request]: Directly mesh via BLE

### DISCUSSION THREAD

**Discussion by EternityForest** [meshtastic/firmware#3829] [Ideas]
### Platform

NRF52, ESP32

### Description

It seems like the protocol itself would run just as well via Bluetooth as it would via LoRa, and modern BLE range can be really good.

Meshing directly via Bluetooth would solve several big issues.  For one thing it would let a group share a single Meshtastic node, sending and receiving messages from their phones, without anyone needing to have settings access to the node.

In an emergency situation, I imagine that could be pretty valuable.   Campgrounds outside of cell range could have permanently installed solar Meshtastic nodes that any guest could use with no setup effort needed.

It would also allow for very low cost/low power/tiny sensors along the lines of AirTags.   A single LoRa device could cover a whole area of tiny nodes that send their messages once a day.

Also, even though it might not be part of the core mission, it would let people try out, and make real use of, all the same features without buying new hardware.

Bluetooth also has virtually unlimited bandwidth compared to LoRa, and there is extended advertising allowing for additional features.  

Bluetooth packets could have "Virtual modem settings" and be repeated onto LoRa only if the modem settings matched the node's setup, allowing you to keep packets off LoRa if they only need to be short range, so short and long range applications could share the same code.

**Comment by charminULTRA** [comment]
Wouldn't BLE range be significantly shorter? Like we measure BLE range in feet, and LoRa range is typically measured in miles.

**Comment by EternityForest** [comment]
BLE's theoretical range is like 500 feet, and even the practical range is
fairly decent.  Nowhere near LoRa, but enough for a lot of use cases,
especially with a BLE to LoRa repeater gateway.

**Comment by mattdrum** [comment]
I could see this functionality being useful for crossband or cross channel repeaters. Additionally, implementing this for two radios on the same network, one being on a directional antenna for a long distance uplink, and a second one being used for omnidirectional coverage.

**Comment by jrichmancodes** [comment]
I was also thinking bluetooth could have more uses such as Nintendo Streetpass did back in the day. I'd love a system where you can identify other nodes nearby via bluetooth and have some sort of key swap to then be able to message each other just by having the devices close and confirming text on each other's screen.

**Reply by dzid26** [reply]
https://bitchat.free/

**Comment by EternityForest** [comment]
I also wonder about ESP-NOW as an interim solution.

It seems a bit easier to implement than BLE, and their proprietary 802.11LR mode can give a whole kilometer of range with stock hardware as I understand it, and it's included in a lot of chips already.

Plus it has encryption features built in.

**Comment by Links2004** [comment]
Hi,
I started looking in to it and it looks possible.
we can send up to 1650 bytes via BLE 5 extended advertising.

First step is to update NimbleBLE to 2.3.6 to get support for this.

see PR:
https://github.com/meshtastic/firmware/pull/7880

Will do some hacky PoC for ESP32 <--> ESP32 mesh if the update is merged.

Note:
I only have ESP-S3 and ESP-C3 bases hardware for testing.

**Comment by Links2004** [comment]
I have a first working setup between ESP32 C3 super mini boards.

The main problem I currently have is that the firmware only supports one radio, which makes building a bridge very hard.

but ESP32-C3 <--- BLE 5 extended advertising ---> ESP32-C3 is working.
connectiong a phone via BLE at the same time works too.

lora region needs to be configured but has no effacts, the code base is not really setuped for a non lora radio ;)

~16m through 3 walls (indoor): 
<img width="400"  alt="image" src="https://github.com/user-attachments/assets/60e6eb6d-f85e-4487-8c52-12118921750f" />

working:
- RX
- TX
- SNR
- RSSI

not working:
- air utilization (nice for statistics but we do not have any limits to handle, the BLE stack will take care if there are some)
- no sleep states (is this possible with out missing messages?)
- bridge between BLE and Lora

https://github.com/Links2004/meshtastic-firmware/tree/ble_mesh

most importend changes:
https://github.com/Links2004/meshtastic-firmware/blob/ble_mesh/src/mesh/BLEMeshInterface.cpp
https://github.com/Links2004/meshtastic-firmware/blob/ble_mesh/src/mesh/BLEMeshInterface.h

side notes

The antenna desigen on most ESP32-C3 super mini boards is very bad, 
I recommand to desolder the chip antenna and replacing it with a 6.2cm wire, will make the module way more usabel.

Its possible to see and send the BLE messages on a phone too (tested via nRF connect on Android).
A Phone can join the mesh directly with out needing any external hardware  to connect too,
but that requires work on the meshtastic APP side or a new deticated APP.

**Reply by GUVWAF** [reply]
Very interesting!
